web/src/hashchange.js

```diff
@@ -157,6 +159,22 @@ function do_hashchange_normal(from_reload) {
                 show_home_view();
                 return false;
             }
+
+            // Just select the message if the narrow is the same as the current narrow and
+            // we already have the new message we want to select rendered.
+            // TODO: Implement a filter comparison function so that we do it for any two narrows
+            // when calling narrow.activate.
+            if (old_hash && hash_util.is_same_narrow(window.location.hash, old_hash)) {
+                const filter = new Filter(terms);
+                if (filter.has_operator("near")) {
+                    const target_id = Number.parseInt(filter.operands("near")[0], 10);
+                    if (message_lists.current.get(target_id)) {
+                        message_lists.current.select_id(target_id, {then_scroll: true});
```

Comment:
I guess if the `narrow.activate` call had a `then_select_offset` calculated, we might want to additionally scroll to that offset?

Reply:
Since `then_select_offset` is associated with `then_selected_id` which can lead user to a different message than `near` id, I think we should just avoid handling that case here. Most common case of that happening would be during reloads and user navigating back to narrow using browser back button, so we would be right to just let `message_view.show` handle that case.